controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,246 @@
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// User set gatekeeper.spec.audit.auditFromCache to Automatic, this reconcile function
+// collect requested constraints and add found kinds which is used in constraint to config.spec.sync.syncOnly
+func (r *ConstraintPodStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+	// This is used for RequeueAfter
+	var requeueTime time.Duration
+
+	gatekeeper := &operatorv1alpha1.Gatekeeper{}
+	// Get gatekeeper resource
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: "",
+		Name:      "gatekeeper",
+	}, gatekeeper)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get config or create if not exist
+	config := &v1alpha1.Config{}
+	err = r.Get(ctx, types.NamespacedName{
+		Namespace: r.Namespace,
+		Name:      "config",
+	}, config)
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			config = &v1alpha1.Config{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "config",
+					Namespace: r.Namespace,
+				},
+			}
+
+			createErr := r.Create(ctx, config)
+			if createErr != nil {
+				log.Error(err, "Fail to create the Gatekeeper Config object, will retry.")
+
+				return reconcile.Result{}, createErr
+			}
+
+			log.Info("The Gatekeeper Config object was created")
+		} else {
+			return reconcile.Result{}, err
+		}
+	}
+
+	constraintPodStatus := &v1beta1.ConstraintPodStatus{}
+
+	err = r.Get(ctx, request.NamespacedName, constraintPodStatus)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Cannot find the ConstraintPodStatus")
+
+			err = r.handleDeleteEvent(ctx, request.Name, config)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{}, nil
+		}
+		// Requeue
+		return reconcile.Result{}, err
+	}
+
+	labels := constraintPodStatus.GetLabels()
+
+	constraintKind := labels["internal.gatekeeper.sh/constraint-kind"]
+	constraintName := labels["internal.gatekeeper.sh/constraint-name"]
+
+	constraintGVR := schema.GroupVersionResource{
+		Group:    "constraints.gatekeeper.sh",
+		Version:  "v1beta1",
+		Resource: strings.ToLower(constraintKind),
+	}
+
+	constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, constraintName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("The Constraint was not found", "constraintName:", constraintName)
+
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	constraintMatchKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+	if err != nil {
+		r.Log.V(1).Info("There is no provided kinds in the Contsraint", "constraintName:", constraintName)
+
+		return reconcile.Result{}, nil //nolint:nilerr
```

Comment:
What if there used to be match kinds but not anymore? Don't you need to clean those up?

Reply:
How can we clean up we don't know which match it is?

Reply:
Wouldn't this be the same as a delete case (e.g. `r.handleDeleteEvent(ctx, request.Name, config)`)?

Reply:
you right I will add